Fix #2683, #2684: Implement Cosmetic Filtering and Scriptlet Injection + Adblock & Stats refactor

Brandon-T opened the pull request:
## Security Review
https://github.com/brave/security/issues/787

## Summary of Changes
-  Implements Cosmetic Filtering and Scriptlet Injection
- Update adblock & adblock-stats APIs to not use `Deferred` library
- Updated Search-Engines Open-Search API to not use `Deferred` library

<!-- Enter a ticket number for this PR, create a new one if it is not there yet. -->
This pull request fixes: #2683, fixes #2684

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:
- Test that `ads` on youtube are blocked correctly.
- Test that the `non-video` ads that show up below a video are blocked correctly.

## Screenshots:
Before (`[AD] amazon.ca - Predator Helios Series`):
![image](https://user-images.githubusercontent.com/1530031/157100544-426c743f-dbaa-4f6a-bb6a-888b27d3e8dc.png)

After (No `[AD]`):
![image](https://user-images.githubusercontent.com/1530031/157100589-ef9b18d7-865d-4461-b90a-f1ac734ac605.png)

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue and pull request is assigned to a milestone (should happen at merge time).

iccub reviewed (commented):
Code looks very good. Next steps:
- Security review, tell them the endpoints used are temporary
- Review from someone on the privacy/adblock team, probably Anton
- As a fun optional exercise we can try moving the resource downloader to use the modern Swift's async/await or Combine
- Push devops/Clifton to finish the backend work for new endpoints for the new resources

Putting `blocked` label on it since dummy endpoint are used in code, can't be really merged

Brandon-T committed:
Validate JSON before passing it to the Rust Engine otherwise, the Rust Engine crashes.

Brandon-T committed:
Added proper conversion of a UTF8 Swift String to a CString suitable for the Rust Engine.

thypon reviewed (commented):
Added a couple of comments!

Brandon-T committed:
Execute the script in sandbox + move code to its own file location to not pollute the delegate.

Brandon-T committed:
Add cosmetic filters and scriptlets.

Use async-await and only operate on temp/new engine to avoid race conditions.
Make setUpFiles run in parallel.
Load into original engine first, then download only into new engines.

Update unit tests and all network requests using download manager to async-await.
Updated NTP to use async-await.
Add check inside loop.
Even more performance boost using another temp engine! :o We should probably fix this on the Rust side (Brave-Core) later on.
Updated URLs to be the correct URLs from Brave's server.
Inject script natively.
